src/module.c

```diff
@@ -6544,11 +6551,26 @@ RedisModuleCallReply *RM_Call(RedisModuleCtx *ctx, const char *cmdname, const ch
      *
      * If RM_SetContextUser has set a user, that user is used, otherwise
      * use the attached client's user. If there is no attached client user and no manually
-     * set user, an error will be returned */
+     * set user, an error will be returned.
+     * An internal command should only succeed for an internal connection, AOF,
+     * and master commands. */
     if (flags & REDISMODULE_ARGV_RUN_AS_USER) {
         int acl_errpos;
         int acl_retval;
 
+        // At this point, we already know that the client is not internal.
+        if (c->cmd->flags & CMD_INTERNAL) {
+            if (mustObeyClient(ctx->client)) {
+                goto acl_ok;
```

Comment:
maybe it's a better idea to do this:
1. move the check for internal connections to right after the call to lookupCommand, and just nullify the command if it isn't exposed to the connectoin.
2. patch ACLCheckAllUserCommandPerm to return without a check for internal connections

i think this can simplify both RM_Call and processCommand

Reply:
That makes sense, but we wanted the internal command validation to be conducted only when the `C` flag is used, and to succeed otherwise.
Do you think the validation should always take place, regardless of the `C` flag?

Reply:
are you sure? if it failed the internal connection test, it won't get to re-process the command

Reply:
The command lookup occurs in the `! client_reprocessing_command` bracket. We would need to have two conditions if we want to put it there